### DISCUSSION TITLE
Get a list of all bonding angles after MMFF optimization

### DISCUSSION THREAD

**Discussion by pythonnewbie3** [rdkit/rdkit#4732] [Q&A]
I want to get a list of all bonding angles (including the ones of hydrogen) in a molecule after MMFF optimization. This is what I have so far.

from rdkit.Chem import rdDistGeom
import rdkit.DistanceGeometry as DG
from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import rdMolTransforms
import random
from rdkit.Chem import rdForceFieldHelpers

mol = Chem.MolFromSmiles("CCC")
m2 = Chem.AddHs(mol)
AllChem.Compute2DCoords(mol)
conf=AllChem.MMFFOptimizeMoleculeConfs(m2)
bonds = [(x.GetBeginAtomIdx(), x.GetEndAtomIdx()) for x in m2.GetBonds()]
print(bonds)
x=rdMolTransforms.GetAngleDeg(conf, 0, 1, 2)

But I get:
ValueError: Bad Conformer Id.

Also this piece of code is not providing me with a list.

Alternatively for my needs a sort of "substructure match" scanning the MMFF optimized conformer for a certain angle (for example 102°) would also work.

I'd be very grateful if someone could help me. Thank you in advance.

**Comment by ptosco** [comment]
@pythonnewbie3 
You create propane from SMILES and add Hs, which is fine:
```
mol = Chem.MolFromSmiles("CCC")
m2 = Chem.AddHs(mol)
```
Then you add 2D coordinates to the molecule without hydrogens:
```
AllChem.Compute2DCoords(mol)
```
Then you try to optimize the 3D conformations on `m2`:
```
conf=AllChem.MMFFOptimizeMoleculeConfs(m2)
```
However, you have never generated any 3D conformations on `m2`, hence the `ValueError: Bad Conformer Id` exception.

First you need to add some 3D conformations to `m2`:
```
conf_ids = AllChem.EmbedMultipleConfs(m2, 5)
```
Then optimization will work:
```
AllChem.MMFFOptimizeMoleculeConfs(m2)
[(0, -4.897293384538459),
 (0, -4.897293376747296),
 (0, -4.8972933819919335),
 (0, -4.897293377551854),
 (0, -4.897293385078721)]
```
Obviously propane conformations are all the same, but this is expected.

To enumerate angles you might enumerate all permutations of two bonds, and check those that have a `j` atom in common, the other two atoms being `i` and `k`.
Or you could loop over all atoms and consider those `j` atoms that have at least 2 neighbors and enumerate their permutations `i` and `k`, which is what the MMFF code does to enumerate angles.
I chose the first as I think it is simpler to understand.

This `get_angle_deg()` function will return a `(i, j, k, angle_deg)` tuple for the pair of bonds `b1`, `b2` of conformation `conf` if the bonds form an angle; it will return `None` if they don't:
```
def get_angle_deg(b1, b2, conf):
    if b1.GetIdx() == b2.GetIdx():
        return None
    if (b1.GetBeginAtomIdx() == b2.GetBeginAtomIdx() or
        b1.GetBeginAtomIdx() == b2.GetEndAtomIdx()):
        j = b1.GetBeginAtomIdx()
    elif (b1.GetEndAtomIdx() == b2.GetBeginAtomIdx() or
          b1.GetEndAtomIdx() == b2.GetEndAtomIdx()):
        j = b1.GetEndAtomIdx()
    else:
        return None
    i = b1.GetOtherAtomIdx(j)
    k = b2.GetOtherAtomIdx(j)
    return (i, j, k, AllChem.GetAngleDeg(conf, i, j, k))
```

The following `get_angles(m, conf_id)` function will enumerate all angle values for molecule `m` and conformer ID `conf_id`:
```
def get_angles(m, conf_id):
    angles = []
    if not m.GetNumBonds():
        return angles
    conf = m.GetConformer(conf_id)
    for b1_idx in range(m.GetNumBonds() - 1):
        b1 = m.GetBondWithIdx(b1_idx)
        for b2_idx in range(b1_idx + 1, m.GetNumBonds()):
            b2 = m.GetBondWithIdx(b2_idx)
            angle = get_angle_deg(b1, b2, conf)
            if angle is not None:
                angles.append(angle)
    return angles
```
So you can generate a dictionary of angles for each conformer ID as follows:
```
angles = {conf_id: get_angles(m2, conf_id) for conf_id in conf_ids}
angles
{0: [(0, 1, 2, 111.6587115134136),
  (1, 0, 3, 111.0019775748325),
  (1, 0, 4, 110.27620823385408),
  (1, 0, 5, 111.002003867221),
  (0, 1, 6, 109.4460381009676),
  (0, 1, 7, 109.44596395034401),
  (2, 1, 6, 109.44599555432022),
  (2, 1, 7, 109.44600244242109),
  (1, 2, 8, 111.00202921702959),
  (1, 2, 9, 110.27626412933404),
  (1, 2, 10, 111.00188584701458),
  (3, 0, 4, 108.08454883681048),
  (3, 0, 5, 108.28496007954018),
  (4, 0, 5, 108.0844666955768),
  (6, 1, 7, 107.29656796670182),
  (8, 2, 9, 108.084545952897),
  (8, 2, 10, 108.28494333208572),
  (9, 2, 10, 108.08449798524383)],
 1: [(0, 1, 2, 111.65872881768514),
  (1, 0, 3, 111.00196889120511),
  (1, 0, 4, 111.00196148220365),
  (1, 0, 5, 110.27622252723543),
  (0, 1, 6, 109.44600669662957),
  (0, 1, 7, 109.44598824636071),
  (2, 1, 6, 109.44602694439813),
  (2, 1, 7, 109.44598304696359),
  (1, 2, 8, 110.27622455302193),
  (1, 2, 9, 111.00198676771203),
  (1, 2, 10, 111.0019532944621),
  (3, 0, 4, 108.28499301128242),
  (3, 0, 5, 108.0845319401815),
  (4, 0, 5, 108.08448912783648),
  (6, 1, 7, 107.29654411373625),
  (8, 2, 9, 108.08453463875333),
  (8, 2, 10, 108.08452672561332),
  (9, 2, 10, 108.28494097388123)],
 2: [(0, 1, 2, 111.65874475069775),
  (1, 0, 3, 110.27621264975738),
  (1, 0, 4, 111.00197310241619),
  (1, 0, 5, 111.00198701522217),
  (0, 1, 6, 109.44598955966529),
  (0, 1, 7, 109.44600075776961),
  (2, 1, 6, 109.44598042638586),
  (2, 1, 7, 109.44596261679584),
  (1, 2, 8, 110.27622918076845),
  (1, 2, 9, 111.00196524101496),
  (1, 2, 10, 111.00196426060032),
  (3, 0, 4, 108.08454344782064),
  (3, 0, 5, 108.08447883560139),
  (4, 0, 5, 108.28497104460443),
  (6, 1, 7, 107.2966014393837),
  (8, 2, 9, 108.08449706586309),
  (8, 2, 10, 108.08452771042346),
  (9, 2, 10, 108.28498337405014)],
 3: [(0, 1, 2, 111.6587326298896),
  (1, 0, 3, 111.00195315162962),
  (1, 0, 4, 110.27623627127436),
  (1, 0, 5, 111.0019741935333),
  (0, 1, 6, 109.44603879654608),
  (0, 1, 7, 109.44597816985326),
  (2, 1, 6, 109.44599437618666),
  (2, 1, 7, 109.44598065272974),
  (1, 2, 8, 111.00196424630121),
  (1, 2, 9, 110.27625651996551),
  (1, 2, 10, 111.00193142755555),
  (3, 0, 4, 108.08452923767865),
  (3, 0, 5, 108.28494143357058),
  (4, 0, 5, 108.08453280472662),
  (6, 1, 7, 107.29655346676006),
  (8, 2, 9, 108.0845921336807),
  (8, 2, 10, 108.28493734501808),
  (9, 2, 10, 108.08448622742438)],
 4: [(0, 1, 2, 111.65871043782185),
  (1, 0, 3, 111.00195259170336),
  (1, 0, 4, 110.2762521793855),
  (1, 0, 5, 111.00196114655715),
  (0, 1, 6, 109.44600858952114),
  (0, 1, 7, 109.44599147682989),
  (2, 1, 6, 109.44599946326201),
  (2, 1, 7, 109.44600689803899),
  (1, 2, 8, 111.0019489197288),
  (1, 2, 9, 111.00197051783465),
  (1, 2, 10, 110.27624615268304),
  (3, 0, 4, 108.08453038057849),
  (3, 0, 5, 108.28494420951202),
  (4, 0, 5, 108.08452655587395),
  (6, 1, 7, 107.29656248353162),
  (8, 2, 9, 108.28495712370791),
  (8, 2, 10, 108.08450758865531),
  (9, 2, 10, 108.08453658803656)]}
```

**Reply by pythonnewbie3** [reply]
Thank you sooo much. :-)

**Reply by pythonnewbie3** [reply]
I wrote this pice of code with you help, because I just want bonding angles for one conformer: 

from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import rdMolTransforms
from itertools import combinations
import numpy as np
import itertools
# Read molecule from smiles string
mol = Chem.MolFromSmiles('[C@@H]12[C@@H]3[C@@H]4[C@H]1[C@H]5[C@@H]4[C@@H]3[C@H]52')
mol = Chem.AddHs(mol)
# Get all bonds in the molecule
bonds = [(x.GetBeginAtomIdx(), x.GetEndAtomIdx()) for x in mol.GetBonds()]
# [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5), (5, 0)]

res = [(*a, b[1]) for a, b in combinations(bonds, r=2) if a[1] == b[0]]

# Compute 2D coordinates
AllChem.EmbedMolecule(mol)
AllChem.MMFFOptimizeMolecule(mol)
conf = mol.GetConformer()

h=[]

for x in range(len(res)):
    k=res[x]
    x=k[0]
    y=k[1]
    z=k[2]
    ang=rdMolTransforms.GetAngleDeg(conf, x, y, z)
    ang=str("%.0f" % ang)
    angl=h.append(ang)
print(h)
count=0
for x in h:
    if x=='90':
        count+=1
print(count)

this however only returns 11 bonding angles with 90° in Cuban, which is wrong. What is my code doing wrong?
